library/core/src/main/java/com/google/android/exoplayer2/text/cea/Cea708Decoder.java

```diff
@@ -270,81 +300,113 @@ private void finalizeCurrentPacket() {
   @RequiresNonNull("currentDtvCcPacket")
   private void processCurrentPacket() {
     if (currentDtvCcPacket.currentIndex != (currentDtvCcPacket.packetSize * 2 - 1)) {
-      Log.w(TAG, "DtvCcPacket ended prematurely; size is " + (currentDtvCcPacket.packetSize * 2 - 1)
+      Log.d(TAG, "DtvCcPacket ended prematurely; size is " + (currentDtvCcPacket.packetSize * 2 - 1)
           + ", but current index is " + currentDtvCcPacket.currentIndex + " (sequence number "
           + currentDtvCcPacket.sequenceNumber + "); ignoring packet");
-      return;
+      // This is not invalid packet. The Packect end can happen either through the reception of
+      // bytes equals to (pack_size_code*2-1) or the reception of cc_type = 0x03 (binary 11).
     }
 
-    serviceBlockPacket.reset(currentDtvCcPacket.packetData, currentDtvCcPacket.currentIndex);
-
-    int serviceNumber = serviceBlockPacket.readBits(3);
-    int blockSize = serviceBlockPacket.readBits(5);
-    if (serviceNumber == 7) {
-      // extended service numbers
-      serviceBlockPacket.skipBits(2);
-      serviceNumber = serviceBlockPacket.readBits(6);
-      if (serviceNumber < 7) {
-        Log.w(TAG, "Invalid extended service number: " + serviceNumber);
+    int currOffset = 0;
+    // Caption channel packet can contain more than one service blocks. Handle all the service
+    // blocks in the current packet. If the previous command spread across the paket, that is
+    // accumulated and handled after all the bytes are received
+    while (currOffset < currentDtvCcPacket.currentIndex) {
+      int serviceNumber = (currentDtvCcPacket.packetData[currOffset] & 0xE0) >> 5; //3 bits
```

Comment:
It seems far preferable to leverage `ParsableBitArray` rather than manually bit shifting.

I wonder if it's possible to re-use the `serviceBlockPacket` one everywhere by some cunning position checking (and rename it `allServiceBlocks`?), or failing that to introduce a second one called `allServiceBlocks` to separate the two.

I had a go locally with the first approach (use `serviceBlockPacket` everywhere) by making edits on top of this change - but I ran into a lack of understanding of exactly how the code is meant to behave (see my next comment) which I think affects how feasible that approach will be.

Reply:
Thanks @icbaker. My understanding of `ParsableBitArray` is that it can not be resized after creation. As per CEA-708 spec, the `cc_data()` can be split across multiple caption channel packets (`section 4.4.1.1` - `A DTVCC Caption Channel Packet may continue from one cc_data() structure to the next`). In such case, we need to retain the previous data to interpret the commands correctly. But  ParsableBitArray.reset() clears everything. If we could extend ParsableBitArray with append feature, I think we can still use ParsableBitArray here.   As a simple workaround, I used ArrayList which can be resized easily.